File: src/tinymonitor/metrics/io.py
import time
import psutil
from tinymonitor.metrics.provider import MetricProvider

class IoMetric(MetricProvider):
    def __init__(self, name, config):
        super().__init__(name, config)
        self.last_counters = psutil.disk_io_counters()
        self.last_time = time.time()

    def parse_threshold(self, value, max_value=None):
        if value is None:
            return float('inf')
        
        if isinstance(value, (int, float)):
            return float(value)
            
        if isinstance(value, str):
            value = value.strip().upper()
            if value.endswith('%'):
                if max_value is None:
                    return float('inf') # Cannot calculate percentage without max
                try:
                    percent = float(value[:-1])
                    return max_value * (percent / 100.0)
                except ValueError:
                    return float('inf')
            
            # Handle units
            units = {'K': 1024, 'M': 1024**2, 'G': 1024**3, 'T': 1024**4}
            for unit, multiplier in units.items():
                # Check for MB, GB, etc.
                if value.endswith(f"{unit}B"):
                     try:
                        num_part = value[:-2]
                        return float(num_part) * multiplier
                     except ValueError:
                        pass
                # Check for M, G, etc.
                elif value.endswith(unit):
                    try:
                        num_part = value[:-1]
                        return float(num_part) * multiplier
                    except ValueError:
                        pass
            
            # Try parsing as plain number
            try:
                return float(value)
            except ValueError:
                pass
                
        return float('inf')

    def check(self):
        current_time = time.time()
        current_counters = psutil.disk_io_counters()
        
        if not self.last_counters or not current_counters:
            self.last_counters = current_counters
            self.last_time = current_time
            return []

        time_delta = current_time - self.last_time
        if time_delta <= 0:
            return []

        read_bytes_delta = current_counters.read_bytes - self.last_counters.read_bytes
        write_bytes_delta = current_counters.write_bytes - self.last_counters.write_bytes
        
        # Avoid negative values if counters reset (unlikely but possible)
        if read_bytes_delta < 0: read_bytes_delta = 0
        if write_bytes_delta < 0: write_bytes_delta = 0
        
        read_speed = read_bytes_delta / time_delta
        write_speed = write_bytes_delta / time_delta
        
        self.last_counters = current_counters
        self.last_time = current_time
        
        # Convert to human readable
        def format_bytes(size):
            power = 2**10
            n = 0
            power_labels = {0 : '', 1: 'K', 2: 'M', 3: 'G', 4: 'T'}
            while size > power:
                size /= power
                n += 1
            return f"{size:.1f}{power_labels.get(n, '')}B/s"

        formatted_read = format_bytes(read_speed)
        formatted_write = format_bytes(write_speed)
        
        level = None
        
        # Parse max_speed if present
        max_speed_val = self.config.get("max_speed")
        max_speed = None
        if max_speed_val:
             max_speed = self.parse_threshold(max_speed_val)
             if max_speed == float('inf'):
                 max_speed = None

        # Thresholds
        warning_val = self.config.get("warning")
        critical_val = self.config.get("critical")
        
        warning_threshold = self.parse_threshold(warning_val, max_speed)
        critical_threshold = self.parse_threshold(critical_val, max_speed)
        
        total_speed = read_speed + write_speed
        
        if total_speed >= critical_threshold:
            level = "CRITICAL"
        elif total_speed >= warning_threshold:
            level = "WARNING"
            
        return [("I/O", level, f"R: {formatted_read} W: {formatted_write}")]

File: tinymonitor-go/internal/metrics/io.go
package metrics

import (
	"fmt"
	"math"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/Gu1llaum-3/tinymonitor/internal/config"
	"github.com/Gu1llaum-3/tinymonitor/internal/models"
	"github.com/shirou/gopsutil/v3/disk"
)

// IOCollector monitors disk I/O
type IOCollector struct {
	name         string
	config       config.IOConfig
	lastCounters *disk.IOCountersStat
	lastTime     time.Time
	mu           sync.Mutex
}

// NewIOCollector creates a new I/O collector
func NewIOCollector(cfg config.IOConfig) *IOCollector {
	counters, _ := disk.IOCounters()
	var total *disk.IOCountersStat
	if len(counters) > 0 {
		t := aggregateCounters(counters)
		total = &t
	}

	return &IOCollector{
		name:         "io",
		config:       cfg,
		lastCounters: total,
		lastTime:     time.Now(),
	}
}

func aggregateCounters(counters map[string]disk.IOCountersStat) disk.IOCountersStat {
	var total disk.IOCountersStat
	for _, c := range counters {
		total.ReadBytes += c.ReadBytes
		total.WriteBytes += c.WriteBytes
	}
	return total
}

// Name returns the collector name
func (c *IOCollector) Name() string {
	return c.name
}

// Duration returns the configured duration threshold
func (c *IOCollector) Duration() int {
	return c.config.Duration
}

// parseThreshold parses a threshold value (number, string with unit, or percentage)
func (c *IOCollector) parseThreshold(value interface{}, maxValue *float64) float64 {
	if value == nil {
		return math.Inf(1)
	}

	switch v := value.(type) {
	case float64:
		return v
	case int:
		return float64(v)
	case string:
		return c.parseStringThreshold(v, maxValue)
	}

	return math.Inf(1)
}

func (c *IOCollector) parseStringThreshold(value string, maxValue *float64) float64 {
	value = strings.TrimSpace(strings.ToUpper(value))

	// Handle percentage
	if strings.HasSuffix(value, "%") {
		if maxValue == nil {
			return math.Inf(1)
		}
		percent, err := strconv.ParseFloat(value[:len(value)-1], 64)
		if err != nil {
			return math.Inf(1)
		}
		return *maxValue * (percent / 100.0)
	}

	// Handle units
	units := map[string]float64{
		"K": 1024,
		"M": 1024 * 1024,
		"G": 1024 * 1024 * 1024,
		"T": 1024 * 1024 * 1024 * 1024,
	}

	for unit, multiplier := range units {
		if strings.HasSuffix(value, unit+"B") {
			numPart := value[:len(value)-2]
			num, err := strconv.ParseFloat(numPart, 64)
			if err == nil {
				return num * multiplier
			}
		} else if strings.HasSuffix(value, unit) {
			numPart := value[:len(value)-1]
			num, err := strconv.ParseFloat(numPart, 64)
			if err == nil {
				return num * multiplier
			}
		}
	}

	// Try plain number
	num, err := strconv.ParseFloat(value, 64)
	if err == nil {
		return num
	}

	return math.Inf(1)
}

func formatBytes(size float64) string {
	power := 1024.0
	n := 0
	labels := []string{"", "K", "M", "G", "T"}

	for size > power && n < len(labels)-1 {
		size /= power
		n++
	}

	return fmt.Sprintf("%.1f%sB/s", size, labels[n])
}

// Check executes the I/O check
func (c *IOCollector) Check() []models.MetricResult {
	c.mu.Lock()
	defer c.mu.Unlock()

	currentTime := time.Now()
	counters, err := disk.IOCounters()
	if err != nil || len(counters) == 0 {
		return nil
	}

	currentCounters := aggregateCounters(counters)

	if c.lastCounters == nil {
		c.lastCounters = &currentCounters
		c.lastTime = currentTime
		return nil
	}

	timeDelta := currentTime.Sub(c.lastTime).Seconds()
	if timeDelta <= 0 {
		return nil
	}

	readBytesDelta := float64(currentCounters.ReadBytes - c.lastCounters.ReadBytes)
	writeBytesDelta := float64(currentCounters.WriteBytes - c.lastCounters.WriteBytes)

	if readBytesDelta < 0 {
		readBytesDelta = 0
	}
	if writeBytesDelta < 0 {
		writeBytesDelta = 0
	}

	readSpeed := readBytesDelta / timeDelta
	writeSpeed := writeBytesDelta / timeDelta

	c.lastCounters = &currentCounters
	c.lastTime = currentTime

	formattedRead := formatBytes(readSpeed)
	formattedWrite := formatBytes(writeSpeed)

	var level *models.Severity

	// Parse max_speed if present
	var maxSpeed *float64
	if c.config.MaxSpeed != nil {
		ms := c.parseThreshold(c.config.MaxSpeed, nil)
		if !math.IsInf(ms, 1) {
			maxSpeed = &ms
		}
	}

	warningThreshold := c.parseThreshold(c.config.Warning, maxSpeed)
	criticalThreshold := c.parseThreshold(c.config.Critical, maxSpeed)

	totalSpeed := readSpeed + writeSpeed

	if totalSpeed >= criticalThreshold {
		sev := models.SeverityCritical
		level = &sev
	} else if totalSpeed >= warningThreshold {
		sev := models.SeverityWarning
		level = &sev
	}

	valueStr := fmt.Sprintf("R: %s W: %s", formattedRead, formattedWrite)
	return []models.MetricResult{
		models.NewMetricResult("I/O", level, valueStr),
	}
}
